include/El/optimization/bfgs.hpp

```diff
@@ -0,0 +1,212 @@
+/*
+   Copyright (c) 2009-2016, Ryan H. Lewis
+   All rights reserved.
+
+   This file is part of Elemental and is under the BSD 2-Clause License, 
+   which can be found in the LICENSE file in the root directory, or at 
+   http://opensource.org/licenses/BSD-2-Clause
+*/
+#ifndef EL_OPTIMIZATION_BFGS_HPP
+#define EL_OPTIMIZATION_BFGS_HPP
+#include <utility>
+#include <El.hpp>
+
+namespace El {
+
+template< typename T, typename Function, typename Gradient>
+T zoom( const Function& f, const Gradient& gradient, T f0, const DistMatrix<T>& x0, const DistMatrix<T>& p,
+        T alpha_low, T alpha_high, T c1, T c2){
+    DistMatrix<T> x_j(x0);
+    DistMatrix<T> g2(p.Height(), 1);
+    while (alpha_high - alpha_low > 10*El::limits::Epsilon<T>()){
+        T alpha = (alpha_low + alpha_high)/2.0;
+        x_j = x0;
+        Axpy(alpha, p, x_j);
+        T fval = f(x_j);
+        if(El::Abs(fval)<= -c2*f0){
+            return alpha;
+        }
+        gradient( x_j, g2);
+        auto fval_dash = Dot(p, g2);
+        if(  fval_dash*(alpha_high - alpha_low) >= 0){
+            alpha_high = alpha_low;
+        }
+        alpha_low = alpha;
+    }
+    return (alpha_high+alpha_low/2.0);
+}
+/***
+ * This function approximately minimizes the function min_\alpha f(x0 + alpha*p)
+ * In some sense minimizing this function is "ideal" however, this may take a lot of work, and since
+ * we are using this inside of a larger algorithm on an approximate problem, there is no reaosn to believe
+ * that this is the "best" step. Instead we just hope for a step with "sufficient" decrease, and then we
+ * hope to prove that if {x_i} approaches the minimum that this function converges.
+ * This algorithm attempts to satsify the weak wolf conditions.
+ */
+template< typename T, typename Function, typename Gradient>
+T lineSearch( const Function& f, const Gradient& gradient,
+              const DistMatrix<T>& g, Int D,
+              const DistMatrix<T>& x0, const DistMatrix<T>& p,
+              Int maxIter=100, T c1=1e-4, T c2=0.9){
```

Comment:
`Pow(limits::Epsilon<Real>(),Real(1)/Real(4))` would return the fourth-root of machine epsilon, which is roughly `1-e4` for double-precision. I'm not saying that this is necessarily the right generalization, but it's worth considering whether you want 1e-4 for single precision.